dockstore-webservice/src/main/java/io/dockstore/webservice/languages/JupyterHandler.java

```diff
@@ -134,29 +142,57 @@ public Map<String, SourceFile> processImports(String repositoryId, String conten
+
         // To avoid listing the contents of non-existent directories and generating non-cachable requests
-        // https://github.com/dockstore/dockstore/pull/5329#discussion_r1088120869
-        // we first determine the contents of '/'
-        Set<String> rootNames = new HashSet<>(ObjectUtils.firstNonNull(sourceCodeRepoInterface.listFiles(repositoryId, "/", version.getReference()), List.of()));
+        // (https://github.com/dockstore/dockstore/pull/5329#discussion_r1088120869),
+        // the following code tends to list the files in the directories of interest and scan through them,
+        // rather than probing single paths.
+        Set<String> rootNames = listFiles.apply("/");
 
-        // For each possible REES directory:
+        // Read the REES configuration files
+        // For each possible REES directory, check to see if it contains any REES files
         for (String reesDir: REES_DIRS) {
             // Confirm the directory [probably] exists.
             if ("/".equals(reesDir) || rootNames.contains(reesDir.replace("/", ""))) {
-                // List the files in the directory.
-                List<String> names = sourceCodeRepoInterface.listFiles(repositoryId, reesDir, version.getReference());
-                if (names != null) {
-                    // Check each file in the directory.
-                    for (String name: names) {
-                        // If it's a REES file, read it into a SourceFile and add it to the map.
-                        if (REES_FILES.contains(name)) {
-                            sourceCodeRepoInterface.readFile(repositoryId, version, DescriptorLanguage.FileType.DOCKSTORE_NOTEBOOK_REES, reesDir + name)
-                                .ifPresent(file -> pathsToFiles.put(file.getAbsolutePath(), file));
-                        }
-                    }
-                }
+                listFiles.apply(reesDir).stream()
+                    .filter(REES_FILES::contains)
+                    .forEach(name -> fileReader.accept(reesDir + name, DOCKSTORE_NOTEBOOK_REES));
             }
         }
+
+        // Scan for and read the devcontainer files of highest precedence
+        // Relevant section of devcontainer spec: https://containers.dev/implementors/spec/#devcontainerjson
+        if (rootNames.contains(DOT_DEVCONTAINER_JSON)) {
+            // Read /.devcontainer.json
+            fileReader.accept("/" + DOT_DEVCONTAINER_JSON, DOCKSTORE_NOTEBOOK_DEVCONTAINER);
+
+        } else if (rootNames.contains(DOT_DEVCONTAINER_DIR)) {
+            String path = "/" + DOT_DEVCONTAINER_DIR;
+            Set<String> names = listFiles.apply(path);
+
+            if (names.contains(DEVCONTAINER_JSON)) {
+                // Read /.devcontainer/devcontainer.json
+                fileReader.accept(path + "/" + DEVCONTAINER_JSON, DOCKSTORE_NOTEBOOK_DEVCONTAINER);
+
+            } else {
+                // Scan for and read all files with path /.devcontainer/<folder>/devcontainer.json
```

Comment:
What are the implications of multiple devcontainer.json files? Does GitHub Codespace automatically pick one when you run the notebook, or prompt you?

If it automatically picks one, should we only index the one that it picks, assuming it's deterministic?

Reply:
Here's an example of the link we'll probably load in the new window, pointing at the branch with the two devcontainers:
https://github.com/codespaces/new?hide_repo_select=true&ref=dotdir-folder-devcontainers&repo=dockstore-testing/simple-notebook&skip_quickstart=true&machine=basicLinux32gb&geo=UsWest
You can select the devcontainer configuration.  Interesting thing is, on the branches where there's one devcontainer, selecting "Default project configuration" will use the devcontainer, but on the branch with two devcontainers, selecting "Default project configuration" will use _neither_.  This raises the possibility that maybe you can have another default devcontainer on that branch, too.  I will do an experiment and see if that's true.

Reply:
Experimental results: Github reads and gives you the option to launch with all of the devcontainers with spec-conformant paths.  It does not register devcontainers from other paths in the tree.  So, we should probably grab all of the former.